### ISSUE TITLE
No inference of array literal nested spread types

### ISSUE THREAD

**Issue by mournie** [microsoft/TypeScript#42037]
ORIGINAL:
We have "special" logic for unifying object literals in these cases; the spread syntax interferes with that.

If you post a more representative sample where it's not just spreading for the sake of spreading, we may be able to advise more.

_Originally posted by @RyanCavanaugh in https://github.com/microsoft/TypeScript/issues/30761#issuecomment-481793894_

NEW:
The common pattern of literal array assignment with conditional inclusion of elements is one use case.
```
type t = 'a'
const a:t[] = [
  ...(true?['a']:[]),
]
```
The error message from the above is `Type 'string[]' is not assignable to type '"a"[]'`.

**Comment by DanielRosenwasser** [comment]
I think that issue is not quite related. This seems pretty reasonable, though in the future please follow the template.

```ts
const a: ("foo" | "bar")[] = [
    ...(Math.random() < 0.5 ? ["foo"] : []),
];
```

[Playground Link](https://www.typescriptlang.org/play?#code/FAYw9gdgzgLgBAQwFxwBQCIBmYzrgHznQCMEAndASgG0BdOAXjmuDjbgDovUBZBGABYcyCCABMwAW1SU4AHjgAGDgFY4AfmZYc6eijqUANMFpA)

**Comment by mournie** [comment]
> This seems pretty reasonable, though in the future please follow the template.
Yeah sorry - I used the "reference in new issue" button and it pre-filled the above quote instead of the standard template, so I just went with that.

**Comment by mournie** [comment]
> I think that issue is not quite related.

I guess object spread types *are* inferred in a way analogous to the way in which the array example *isn't*, so here's another example.

```
type a = {one: 1}
type b = 'b'[]

const a:a = {...{one:1}} // work
const b:b = [...['b']] // not work
```

**Comment by RyanCavanaugh** [comment]
Let's track at #45600
